cli/cli_app.go

```diff
@@ -183,54 +189,165 @@ func runCommand(command string, terragruntOptions *options.TerragruntOptions) (f
+			terragruntOptions.InsertTerraformCliArgs(terragruntConfig.RemoteState.ToTerraformInitArgs()...)
+		}
+
+	} else {
+		needsInit, err := needsInit(terragruntOptions, terragruntConfig)
+		if err != nil {
+			return err
+		}
+
+		if needsInit {
+			if err := runTerraformInit(terragruntOptions, terragruntConfig, nil); err != nil {
+				return err
+			}
+		}
+
 	}
+	return shell.RunTerraformCommand(terragruntOptions, terragruntOptions.TerraformCliArgs...)
+}
 
-	if err := downloadModules(terragruntOptions); err != nil {
-		return err
+// Determines if 'terraform init' needs to be executed
+func needsInit(terragruntOptions *options.TerragruntOptions, terragruntConfig *config.TerragruntConfig) (bool, error) {
+
+	if util.ListContainsElement(TERRAFORM_COMMANDS_THAT_DO_NOT_NEED_INIT, firstArg(terragruntOptions.TerraformCliArgs)) {
+		return false, nil
+	}
+
+	modulesNeedsInit, err := modulesNeedInit(terragruntOptions)
+	if err != nil {
+		return false, err
+	}
+	if modulesNeedsInit {
+		return true, nil
 	}
 
-	if conf.RemoteState != nil {
-		if err := configureRemoteState(conf.RemoteState, terragruntOptions); err != nil {
+	remoteStateNeedsInit, err := remoteStateNeedsInit(terragruntConfig.RemoteState, terragruntOptions)
+	if err != nil {
+		return false, err
+	}
+	if remoteStateNeedsInit {
+		return true, nil
+	}
+	return false, nil
+
+}
+
+// Runs the terraform init command to perform what is referred to as Auto-Init in the README.md.
+// This is intended to be run when the user runs another terragrunt command (e.g. 'terragrunt apply'),
+// but terragrunt determines that 'terraform init' needs to be called pror to runing
+// the respective terraform command (e.g. 'terraform apply')
+//
+// The terragruntOptions are assumed to be the options for running the original terragrunt command.
+//
+// If terraformSource is specified, then arguments to download the terraform source will be appended to the init command.
+//
+// This method will return an error and NOT run terraform init if the user has disabled Auto-Init
+func runTerraformInit(terragruntOptions *options.TerragruntOptions, terragruntConfig *config.TerragruntConfig, terraformSource *TerraformSource) error {
+
+	// Prevent Auto-Init if the user has disabled it
+	if firstArg(terragruntOptions.TerraformCliArgs) != CMD_INIT && !terragruntOptions.AutoInit {
+		return errors.WithStackTrace(InitNeededButDisabled("Cannot continue because init is needed, but Auto-Init is disabled.  You must run 'terragrunt init' manually."))
+	}
+
+	// Need to clone the terragruntOptions, so the TerraformCliArgs can be configured to run the init command
+	initOptions := terragruntOptions.Clone(terragruntOptions.TerragruntConfigPath)
+	initOptions.TerraformCliArgs = []string{CMD_INIT}
+	initOptions.WorkingDir = terragruntOptions.WorkingDir
+
+	// Only add the arguments to download source if terraformSource was specified
+	downloadSource := terraformSource != nil
+	if downloadSource {
```

Comment:
Shy the extra variable instead of just checking `terraformSource != nil` directly in the if-statement?